src/main/kotlin/com/openlattice/data/DataGraphService.kt

```diff
@@ -465,6 +475,35 @@ open class DataGraphService(
         return null
     }
 
+    private fun checkAssociationEntityTypes(associations: Set<DataEdgeKey>) {
+        associations.forEach {
```

Comment:
Looping over the days keys twice seems unnecessarily expensive. On top of that the API calls we are making do not use a projection yet (that's a separate ticket) so they loading they entire entity set over and over again over the network (quite expensive). Accumulate a Map<UUID, Map<UUID, Set<UUID>>> (edge entity set id -> src entity set id -> dst entity set ids) then load just the distinct entity sets, grab their entity types like you are currently doing, and then follow your current approach except that you are only evaluating distinct entity set combinations (which will tend to be much smaller) then total items submitted.